### ISSUE TITLE
Add a module for making contract calls

### ISSUE THREAD

**Issue by masonforest** [mana-ethereum/ethereumex#49]
Currently we make contract calls by encoding ABI manually.

    abi_encoded_data = ABI.encode("transferFrom(address,address,uint)",[from_address, to_address, token_id])
    contract_address = "0x123" |> String.slice(2..-1) |> Base.decode16(case: :mixed)

    transaction_data = %Blockchain.Transaction{
        data: abi_encoded_data,
        gas_limit: 100_000,
        gas_price: 16_000_000_000,
        init: <<>>,
        nonce: 5,
        to: contract_address,
        value: 0
    }
    |> Blockchain.Transaction.Signature.sign_transaction(private_key)
    |> Blockchain.Transaction.serialize()
    |> ExRLP.encode()
    |> Base.encode16(case: :lower)

    Ethereumex.HttpClient.eth_send_raw_transaction("0x" <> transaction_data)

I think we could make a cleaner interface.

Here’s a sketch of the API I propose:

    abi = File.read!(“ERC20.abi”)
    contract_address = "0x123" |> String.slice(2..-1) |> Base.decode16(case: :mixed)
    contract = Ethereumex.Contract.new(%{abi: abi, address: contract_address})
    client = Ethereumex.Client.new(%{gas_limit: 100_000, gas_price: 16_000_000_000})
    account =  Ethereumex.Client.new(%{private_key: private_key, nonce: 5})

    {:ok, account, transaction_hash} = contract
      |> Ethereumex.call(:transfer_to, from_address, to_address, amount)
      |> client.send(account)

This would take some refactoring but I think it would make calling contracts in client code much cleaner.

It’s possible that this is a higher level abstraction than this library is meant to provide. If that’s the case we/I can pull it into a new package. `web3ex`?

To recap:

1. Would anyone else benefit from the API I suggested?
2. Should this live in this repo or a new one?

If anyone has any thoughts feel free to comment here or join the [mana gitter](https://gitter.im/mana-ethereum/mana).

**Comment by hayesgm** [comment]
Makes sense to me, FYI, here's some code I had lying around relating to pulling ABI data from JSON files:

```elixir

defmodule ContractABI do

  defmodule ABI do
    @type type :: String.t()
    @type state_mutability :: nil | :view | :nonpayable
    @type abi_type :: nil | :event | :function | :constructor | :fallback

    @type io_data_type :: %{
            type: type,
            name: String.t(),
            indexed: boolean()
          }

    @type t :: %{
            constant: boolean(),
            inputs: list(io_data_type()),
            name: String.t(),
            outputs: list(io_data_type()),
            payable: boolean(),
            state_mutability: state_mutability(),
            abi_type: abi_type()
          }

    defstruct constant: false,
              inputs: [],
              name: "",
              outputs: [],
              payable: false,
              state_mutability: nil,
              abi_type: nil
  end

  @spec parse_contract_abi(String.t()) :: %{String.t() => {list(ABI.t()), list(base_abi())}}
  def parse_contract_abi(abi_json) when is_binary(abi_json) do
    json = Poison.decode!(abi_json)

    for {contract_name, abi_list} <- json, into: %{} do
      contract_abi =
        for abi_map <- abi_list do
          %ABI{
            constant: abi_map["constant"] == true,
            name: abi_map["name"],
            inputs: (abi_map["inputs"] || []) |> Enum.map(&parse_data_type/1),
            outputs: (abi_map["outputs"] || []) |> Enum.map(&parse_data_type/1),
            payable: abi_map["payable"] == true,
            state_mutability: get_state_mutability(abi_map["stateMutability"]),
            abi_type: get_abi_type(abi_map["type"])
          }
        end

      {contract_name, {contract_abi, abi_list}}
    end
  end

  def parse_contract_abi(map) when is_map(map), do: map
  def parse_contract_abi(els), do: raise("Unknown ABI data: #{inspect(els)}")

  @spec parse_data_type(%{}) :: ABI.io_data_type()
  defp parse_data_type(data_type) do
    %{
      name: data_type["name"],
      type: data_type["type"],
      indexed: data_type["indexed"] == true
    }
  end

  @spec get_state_mutability(String.t()) :: ABI.state_mutability()
  defp get_state_mutability(state_mutability) do
    case state_mutability do
      "view" -> :view
      "nonpayable" -> :nonpayable
      _ -> nil
    end
  end

  @spec get_abi_type(String.t()) :: ABI.abi_type()
  defp get_abi_type(type) do
    case type do
      "function" -> :function
      "event" -> :event
      "fallback" -> :fallback
      "constructor" -> :constructor
      _ -> nil
    end
  end
end
```

**Comment by hswick** [comment]
FWIW I have most, if not all of this implemented here https://github.com/hswick/exw3
If there is any reason that library wouldn't suit your purpose I would be happy to hear it.

**Comment by hayesgm** [comment]
I do personally like the idea of having a raw library (Ethereumex) and a higher-level library (ExW3). What do you think @masonforest ?

**Comment by masonforest** [comment]
Yeah, I'm happy having two libraries 👍 

I hadn't seen `exw3`! Just what I was looking for. Thanks @hswick !
